README.md
# Draw Note

[![](https://img.shields.io/github/stars/zhaouv/vscode-markdown-draw.svg)](https://github.com/zhaouv/vscode-markdown-draw/stargazers) [![](https://img.shields.io/github/forks/zhaouv/vscode-markdown-draw.svg)](https://github.com/zhaouv/vscode-markdown-draw/network/members) [![](https://img.shields.io/github/issues/zhaouv/vscode-markdown-draw.svg)](https://github.com/zhaouv/vscode-markdown-draw/issues) [![](https://img.shields.io/github/license/zhaouv/vscode-markdown-draw.svg)](https://github.com/zhaouv/vscode-markdown-draw/blob/master/LICENSE) [![](https://vsmarketplacebadge.apphb.com/version/zhaouv.vscode-markdown-draw.svg)](https://marketplace.visualstudio.com/items?itemName=zhaouv.vscode-markdown-draw)

Features **shortcut bindings** / **saving to SVG file** / **auto folding** added

**This project is in the prototype stage !**

Draw svg quick-responsively with a mouse or pen,  
Designed for note-taking, formulas drafting

OCR to latex supported now (by [Customize-Buttons and myscript API](#Convert-formulas-to-latex))

![](img_md/mainuidemo.png)

## Usage

open markdown file, right click at some line,  
`Edit/Create current line as SVG`

or press `F1`, type `svg`, click the command

## Prototype Stage

> Compatibility is not guaranteed in prototype stage

Todo:

+ [x] fold the svg automatically
+ [ ] adjust behavior of pens
+ [ ] fix the bug of polygon

Welcome [star](https://github.com/zhaouv/vscode-markdown-draw/stargazers) to support the project or [issue](https://github.com/zhaouv/vscode-markdown-draw/issues) to provide ideas

## Customize Buttons

I have not decided the interface. In this stage, it could work this way.

Step 1. You should change the following code(icon/title/onclick), run it in chrome console. It prints one line of string.

```js
console.log(JSON.stringify((()=>{

    // find a icon from https://fontawesome.com/v5.8.2/icons
    const icon = 'beer' 
    const title = 'a new button'
    const addGap = true
    const onclick = ()=>{
        drawAPI.unstable.setTextContent('a custom button works')
    }

    document.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd',(addGap?"<span class='svgiconspliter'></span>":"")+"<div class='svg-btn fa fa-"+icon+"' title='"+title+"'><span></span></div>");
    document.querySelector('div.svg-operate > :last-child').onclick=onclick

}).toString().slice(7,-2)))
```

`"    // find a icon from https://fontawesome.com/v5.8.2/icons\n    const icon = 'beer' \n    const title = 'a new button'\n    const addGap = true\n    const onclick = ()=>{\n        drawAPI.unstable.setTextContent('a custom button works')\n    }\n\n    document.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd',(addGap?\"<span class='svgiconspliter'></span>\":\"\")+\"<div class='svg-btn fa fa-\"+icon+\"' title='\"+title+\"'><span></span></div>\");\n    document.querySelector('div.svg-operate > :last-child').onclick=onclick\n"`

Step 2. Push the string into your settings.json

```json
    "markdown-draw.customized-buttons": [
        {
            "type": "script",
            "version": "0.1.2",
            "function": "    // find a icon from https://fontawesome.com/v5.8.2/icons\n    const icon = 'beer' \n    const title = 'a new button'\n    const addGap = true\n    const onclick = ()=>{\n        drawAPI.unstable.setTextContent('a custom button works')\n    }\n\n    document.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd',(addGap?\"<span class='svgiconspliter'></span>\":\"\")+\"<div class='svg-btn fa fa-\"+icon+\"' title='\"+title+\"'><span></span></div>\");\n    document.querySelector('div.svg-operate > :last-child').onclick=onclick\n"
        }
    ]
```

### Convert formulas to latex

<details open>
<summary>myscript API</summary>

Recognitize handwritten text to latex using API of myscript.com  
which declares `MyScript has been working on this matter for over two decades now and there is no question that our technology is ahead of Microsoft Ink Recognizer. ` at their [article](https://medium.com/@myscript/microsoft-ink-recognizer-an-opportunity-for-myscript-9e55fe45afae).  
Create a account at https://developer.myscript.com/getting-started/web to get token.  
It is free for 2000 calls/month.  

Paste your token on `PasteYourTokenHere`, and push this into setting.json following the previous section [Customize Buttons](#Customize-Buttons). The json generated from [this script](https://github.com/zhaouv/vscode-markdown-draw/blob/master/buttons_demo/htr_to_latex_myscriptapi.js)

```json
{"type":"script","version":"0.1.2","function":"var token = {\n applicationKey: 'PasteYourTokenHere',\n hmacKey: 'PasteYourTokenHere',\n}\nvar addGap = true\n\nvar iink_cdn = 'https://cdn.jsdelivr.net/npm/iink-js@1.4.5/dist/iink.min.js'\n// var iink_cdn = https://unpkg.com/iink-js@1.4.5/dist/iink.min.js\n// var iink_cdn = https://myscript.github.io/iinkJS/dist/iink.min.js\n\nvar icon = 'square-root-alt'\nvar title = 'Recognize to latex'\n\ndocument.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd', (addGap ? \"<span class='svgiconspliter'></span>\" : \"\") + \"<div class='svg-btn fa fa-\" + icon + \"' title='\" + title + \"'><span></span></div>\");\nvar btnElement = document.querySelector('div.svg-operate > :last-child')\nbtnElement.onclick = ()=>{drawAPI.unstable.setTextContent('loading script')}\n\nvar svg = drawAPI.unstable.getSVGElement()\n\nvar getStrokeGroups = () => {\n let strokes = []\n for (const item of svg.children) {\n if (item.nodeName === 'path') {\n let points = item\n .getAttributeNS(null, \"d\")\n .split(/M |L /g)\n .slice(1)\n .map(item => {\n return {\n x: parseFloat(item.split(\",\")[0]),\n y: parseFloat(item.split(\",\")[1])\n };\n })\n strokes.push({ x: points.map(v => v.x), y: points.map(v => v.y) })\n }\n }\n console.log(strokes);\n return [{ \"penStyle\": null, \"strokes\": strokes }]\n}\n\nvar s1 = document.createElement('script')\ns1.setAttribute('nonce', drawAPI.unstable.nonce())\ns1.src = iink_cdn\ns1.onload = () => {\n\n const onclick = () => {\n drawAPI.unstable.setTextContent('calling the API')\n // Creating a recognizer\n const iinkRecognizer = iink.DefaultBehaviors.recognizerList.find(x => {\n const infos = x.getInfo();\n return infos.protocol === 'REST';\n });\n\n // Creating a empty model\n const model = iink.InkModel.createModel();\n // Filling the model with the stroke groups\n model.strokeGroups = getStrokeGroups();\n\n // Creating a recognizer context with the configuration attached\n const recognizerContext = iink.RecognizerContext.createEmptyRecognizerContext({\n configuration: iink.DefaultConfiguration\n });\n\n recognizerContext.editor.configuration.recognitionParams = {\n type: 'MATH',\n protocol: 'REST',\n server: {\n scheme: 'https',\n // host: 'webdemoapi.myscript.com',\n host: 'cloud.myscript.com',\n applicationKey: token.applicationKey,\n hmacKey: token.hmacKey\n },\n iink: {\n math: {\n mimeTypes: [\n 'application/x-latex',\n ],\n }\n }\n }\n\n // Assigning a theme to the document\n recognizerContext.editor.theme = iink.DefaultTheme;\n\n // Defining the behaviour on recognition result\n const recognitionCallback = (err, x) => {\n if (!err) {\n Object.entries(x.exports)\n .forEach(([mimeType, exportValue]) => {\n\n let latex = x.exports[mimeType]\n let content = '\\n$$'+latex.trim()+'$$ '+' '\n drawAPI.unstable.setTextContent('')\n drawAPI.unstable.editCurrentLine({\n control: 0,\n text: content\n })\n });\n }\n };\n\n // Triggering the recognition\n iinkRecognizer.export_(recognizerContext, model)\n .then((values) => {\n values.forEach((value) => {\n recognitionCallback(undefined, value);\n });\n })\n .catch(err => recognitionCallback(err, undefined));\n }\n\n btnElement.onclick = onclick\n}\ndocument.body.appendChild(s1)\n"}
```

Actually it is HTR not OCR.  

</details>

<details>
<summary>mathpix API (click to expand)</summary>

The mathpix API is a popular choice for OCR to latex.  
Its API it free for 1000/month, but you have to provide a card. 
They charge a one-time non-refundable setup fee of $1.  

The following script also supports drag and paste image to latex by mathpix API.  

And I am considering providing another independent extension to convert the clipbord-picture to latex.  
Which need not to open a webview panel and provides a command and is able to bind keys.  
(It will be <https://github.com/zhaouv/vscode-paste-to-latex-mathpix.git> if it becomes existence.)


```json
{"type":"script","version":"0.1.2","function":"var token = { app_id: 'PasteYourTokenHere', app_key: 'PasteYourTokenHere' }\nvar addGap = true\nvar icon = 'square-root-alt'\nvar title = 'Recognize to latex'\n\ndocument.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd', (addGap ? \"<span class='svgiconspliter'></span>\" : \"\") + \"<div class='svg-btn fa fa-\" + icon + \"' title='\" + title + \"'><span></span></div>\");\nvar btnElement = document.querySelector('div.svg-operate > :last-child')\nbtnElement.onclick = () => {\n\n drawAPI.unstable.getPNG((dataURL) => {\n drawAPI.unstable.setTextContent('calling the API')\n xhrPost(dataURL, (err,ret)=>{\n console.log(err,ret)\n let latex = JSON.parse(ret)['latex']\n let content = '\\n$$'+latex.trim()+'$$ '+' '\n drawAPI.unstable.setTextContent('')\n drawAPI.unstable.editCurrentLine({\n control: 0,\n text: content\n })\n })\n })\n}\n\nfunction xhrPost(dataURL, callback) {\n var xhr = new XMLHttpRequest();\n xhr.onreadystatechange = function () {\n if (xhr.readyState == 4) {\n if ((xhr.status >= 200 && xhr.status < 300) || xhr.status == 304) {\n callback(null, xhr.responseText);\n } else {\n callback([xhr.status, xhr.responseText], null);\n }\n }\n }\n xhr.open('post', 'https://api.mathpix.com/v3/latex');\n xhr.setRequestHeader('app_id', token.app_id)\n xhr.setRequestHeader('app_key', token.app_key)\n xhr.setRequestHeader('Content-type', 'application/json')\n xhr.send(JSON.stringify({ 'url': dataURL }));\n}\n\n// drag and paste image\nfunction convertImage(dataURL) {\n drawAPI.unstable.setTextContent('calling the API to convert image')\n xhrPost(dataURL, (err,ret)=>{\n console.log(err,ret)\n let latex = JSON.parse(ret)['latex']\n let content = '\\n$$'+latex.trim()+'$$ '+' '\n drawAPI.unstable.setTextContent('')\n drawAPI.unstable.editCurrentLine({\n control: 0,\n text: content\n })\n })\n\n}\nfunction getImage(items, cb) {\n var file = null;\n if (items && items.length) {\n for (var i = 0; i < items.length; i++) {\n if (items[i].type.indexOf('image') !== -1) {\n file = items[i].getAsFile();\n break;\n }\n }\n }\n // console.log(file);\n if (file) {\n var reader = new FileReader()\n reader.onload = function (event) {\n cb(event.target.result);\n }\n reader.readAsDataURL(file);\n }\n}\nvar bindElement = document.body\nbindElement.addEventListener('paste', function (event) {\n var items = event.clipboardData?.items;\n getImage(items,convertImage);\n});\nbindElement.ondragover = function (ev) {\n ev.preventDefault();\n}\nbindElement.ondrop = function (ev) {\n ev.preventDefault();\n var items = ev.dataTransfer.items;\n getImage(items,convertImage);\n}\n"}
```
Paste your token on `PasteYourTokenHere`, and push this into setting.json following the previous section [Customize Buttons](#Customize-Buttons). The json generated from [this script](https://github.com/zhaouv/vscode-markdown-draw/blob/master/buttons_demo/ocr_to_latex_mathpix.js)

</details>

## webview-keybindings

default keybind is
```json
    "markdown-draw.webview-keybindings": [
        {
            "key": "ctrl+s",
            "command": "save inline"
        },
        {
            "key": "ctrl+f",
            "command": "save to .svg file"
        },
        {
            "key": "ctrl+z",
            "command": "undo"
        },
        {
            "key": "ctrl+y",
            "command": "redo"
        },
        {
            "key": "ctrl+l",
            "command": "Recognize to latex"
        },
        {
            "key": "f2",
            "command": "line"
        },
        {
            "key": "f4",
            "command": "pen"
        },
        {
            "key": "f6",
            "command": "eraser"
        },
        {
            "key": "f7",
            "command": "rect"
        },
        {
            "key": "f8",
            "command": "circle"
        },
        {
            "key": "f9",
            "command": "select"
        },
        {
            "key": "f10",
            "command": "clear board"
        }
    ]
```

Only support `ctrl+<some char>` `f<1~12>` now (just because I am lazy, PR is willcome to support `ctrl+shift+<some char>` and etc.).

If you want to cancel a keybinding, just move it out of the list or change it as
```json
        {
            "key": "f2",
            "command": ""
        },
```

When you press the key. A click will send to the first title-match button.

## License

[Apache-2.0](./LICENSE)

board adapted from [syfxlin/graffiti-board](https://github.com/syfxlin/graffiti-board/tree/5945b126c945073eced5e6eb78658bc2a7375881) Apache-2.0  
> with lots of changes

board/Font-Awesome-5-8-2-all-min.css from Font Awesome Free 5.8.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License)  
> replace webfonts/fa-solid-900.woff2 by base64

CHANGELOG.md
# Change Log

## 0.1.5 (2022-06-05)

Shortcut bindings of webview (reference: fork of [joseaeltala](https://github.com/joseaeltala/vscode-markdown-draw/commit/b24a96ac7710a7177461bf7056152c79a81bd97d) and [JL-MORENOCID](https://github.com/zhaouv/vscode-markdown-draw/compare/master...JL-MORENOCID:master))

Collapse svg input element and add toggle button [PR#4](https://github.com/zhaouv/vscode-markdown-draw/pull/4). Thanks [hall](https://github.com/hall).

Allow storing svg in external file [PR#3](https://github.com/zhaouv/vscode-markdown-draw/pull/3). Thanks [hall](https://github.com/hall) again!

Auto folding (can be disabled in settings)

## 0.1.4 (2021-03-01)

HTR to latex (by mathpix API)  
Drag and paste image to latex (by mathpix API)  

## 0.1.3 (2020-12-08)

icon

## 0.1.2 (2020-12-06)

HTR to latex (by myscript API)  
Customize buttons  
CSP script nonce

## 0.1.1 (2020-11-29)

Initial Release

A prototype which is just able to draw.  
Lots of bug and todo.

## 0.1.0 (2020-11-17)

Initial Project

board/page.js
const drawAPI = {
  unstable: {
    nonce: () => 'ToBeReplacedByRandomToken',
    editCurrentLine(args) {
      console.log({
        ...args,
        command: 'editCurrentLine',
      });
    },
  },
}
window.drawAPI = drawAPI

const lineContentInput = document.querySelector('input[type=text]');

function setContent(message) {
  lineContentInput.value = message.content;
  var svg = document.getElementById('svg');
  var div_container = document.querySelector('div.container')
  // 不准确且不安全的检查, 先这么写
  if (message.content.startsWith('<svg id="svg"')) {
    svg.outerHTML = message.content
      .replace(/<svg id="svg" (?:viewbox|style)=["'][^>]+["']/, '<svg id="svg"')
  } else {
    svg.outerHTML = '<svg id="svg"></svg>'
  }
  // 代价很大, 临时先这么写
  div_container.outerHTML = div_container.outerHTML
  window.paints = []
  initPaint("svg");
}

// Handle the message inside the webview
window.addEventListener('message', event => {

  const message = event.data // The JSON data our extension sent
    || event.detail; // for debug in chrome

  switch (message.command) {
    case 'currentLine':
      setContent(message);
      break;
  }
});

document.querySelector('#text-nextline').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 'nextline',
    text: lineContentInput.value
  })
};
document.querySelector('#text-change-stay').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 'stay',
    text: lineContentInput.value
  })
};
document.querySelector('#text-change-nextline').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 'nextline',
    text: lineContentInput.value
  })
};
(function () {
  if (typeof acquireVsCodeApi !== 'undefined') {
    const vscode = acquireVsCodeApi();
    drawAPI.unstable.editCurrentLine = (args) => {
      vscode.postMessage({
        ...args,
        command: 'editCurrentLine',
      })
    }
    vscode.postMessage({ command: 'requestCurrentLine' })
  }
}());

board/webview.js
if (typeof require !== 'undefined') {
  var { initPaint } = require('./main')
} else {
  var { initPaint } = exports
}

const exposedFunctions = initPaint("svg");

const svgElement = document.getElementById('svg')
const lineContentInput = document.querySelector('input[type=text]');

const drawAPI = {
  unstable: {
    nonce: () => 'ToBeReplacedByRandomToken',
    /**
     * 
     * @param {String} text text
     * @param {Number} control moving number of the cursor
     */
    editCurrentLine({ text, control, ...rest }) {
      console.log({
        text,
        control,
        file:!!rest.file,
        command: 'editCurrentLine',
      });
    },
    readSVGFileContent(file) {
      console.log({
        file,
        command: 'readSVGFile',
      })
    },
    getSVGElement: () => svgElement,
    /**
     * 
     * @param {(dataURL:String)=>undefined} cb callback
     */
    getPNG(cb) {
      var svg = drawAPI.unstable.getSVGElement()
      var { x, y, width, height } = svg.getBBox();
      var b = 10
      var b2 = 2 * b
      var wb = width + x + b
      var hb = height + x + b
      var w = width + b2
      var h = height + b2
      var data = svg.outerHTML.replace('<svg id="svg">',
        `<svg id="svgpng" viewbox="${x - b},${y - b},${w},${h}" style="height:${h}" xmlns="http://www.w3.org/2000/svg">`);
      var r = window.devicePixelRatio || 1;
      var setsize = (ele, ww, hh) => {
        ele.style.width = ww + 'px'
        ele.style.height = hh + 'px'
        ele.setAttribute('width', ww * r)
        ele.setAttribute('height', hh * r)
      }
      var can = document.createElement('canvas')
      setsize(can, w, h)
      var ctx = can.getContext('2d')
      var can2 = document.createElement('canvas')
      setsize(can2, wb, hb)
      var ctx2 = can2.getContext('2d')
      ctx2.scale(r, r)
      var img = new Image();
      setsize(img, wb, hb)
      img.onload = function () {
        ctx2.drawImage(img, 0, 0)
        ctx.putImageData(ctx2.getImageData((x - b) * r, (y - b) * r, w * r, h * r), 0, 0);
        cb(can.toDataURL())
        // document.body.append(can)
      };
      img.src = 'data:image/svg+xml;base64,' + btoa(data);
    },
    reRegisterSVG() {
      exposedFunctions.reInit();
    },
    setTextContent(content) {
      lineContentInput.value = content;
    },
    setSVGContent(content) {
      // 可能有问题的替换
      svgElement.innerHTML = content
        .replace(/<svg id="svg"[^>]*>/, '')
        .replace(/<\/svg>/, '')
      drawAPI.unstable.reRegisterSVG()
    },
    setContent(content) {
      drawAPI.unstable.setTextContent(content)
      document.querySelector("#svg-clean")?.dispatchEvent(new Event('click'))
      let match;
      // 不准确的检查, 先这么写
      if (content.startsWith('<svg id="svg"')) {
        drawAPI.unstable.setSVGContent(content)
      }
      else if (match = /!\[.*\]\((.*\.svg)\)/.exec(content)) {
        drawAPI.unstable.readSVGFileContent(match[1])
      }
    },
    custom(content) {
      console.log(content);
      if (content.operate) {
        content.operate.forEach(drawAPI.unstable.customOperate);
      }
    },
    customOperate(operate) {
      console.log(operate);
      if (operate.type === 'script') {
        let func = new Function(operate.function)
        func()
      }
    },
    keybindings: {},
    updatekeybindings(webviewKeybindings) {
      webviewKeybindings.forEach(v=>{
        if (v.key) {
          drawAPI.unstable.keybindings[v.key.toLowerCase()]=v.command
        }
      })
    },
    ShortCuts(e){
      let evtobj = window.event? event : e
      let key = evtobj.key
      if (evtobj.ctrlKey) {
        key='ctrl+'+key
      }
      let title = drawAPI.unstable.keybindings[key.toLowerCase()]
      console.log(evtobj,key.toLowerCase(),title)
      if (title) {
        let buttons = document.querySelectorAll(".svg-btn")
        for (const button of buttons) {
          if(button.getAttribute('title') == title){
            button.dispatchEvent(new Event('click'))
            return
          }
        }
      }
    },
  },
}
window.drawAPI = drawAPI

// Handle the message inside the webview
window.addEventListener('message', event => {

  const message = event.data // The JSON data our extension sent
    || event.detail; // for debug in chrome

  switch (message.command) {
    case 'currentLine':
      drawAPI.unstable.setContent(message.content);
      break;
    case 'custom':
      drawAPI.unstable.custom(message.content);
      break;
    case 'readSVGFile':
      drawAPI.unstable.setSVGContent(message.content);
      break;
  }
});

document.querySelector('#text-nextline').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 1,
    text: lineContentInput.value
  })
};
document.querySelector('#text-change-stay').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 0,
    text: lineContentInput.value
  })
};
document.querySelector('#text-change-nextline').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 1,
    text: lineContentInput.value
  })
};
document.querySelector('#text-save-file').onclick = function () {
  drawAPI.unstable.editCurrentLine({
    control: 0,
    text: lineContentInput.value,
    file: true
  })
};
document.onkeydown = drawAPI.unstable.ShortCuts;

(function () {
  if (typeof acquireVsCodeApi !== 'undefined') {
    const vscode = acquireVsCodeApi();
    drawAPI.unstable.editCurrentLine = ({ text, control, ...rest }) => {
      vscode.postMessage({
        text,
        control,
        file:!!rest.file,
        command: 'editCurrentLine',
      })
    }
    drawAPI.unstable.readSVGFileContent = (file) => {
      vscode.postMessage({
        file,
        command: 'readSVGFile',
      })
    }
    vscode.postMessage({ command: 'requestCurrentLine' })
    vscode.postMessage({ command: 'requestCustom' })
  }
}());

buttons_demo/htr_to_latex_myscriptapi.js
/* 
Recognitize handwritten text to latex using api of myscript.com
Create a account at https://developer.myscript.com/getting-started/web to get token
It is free for 2000 calls/month
*/
var fff=()=>{

var token = {
    applicationKey: 'PasteYourTokenHere',
    hmacKey: 'PasteYourTokenHere',
}
var addGap = true

var iink_cdn = 'https://cdn.jsdelivr.net/npm/iink-js@1.4.5/dist/iink.min.js'
// var iink_cdn = https://unpkg.com/iink-js@1.4.5/dist/iink.min.js
// var iink_cdn = https://myscript.github.io/iinkJS/dist/iink.min.js

var icon = 'square-root-alt'
var title = 'Recognize to latex'

document.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd', (addGap ? "<span class='svgiconspliter'></span>" : "") + "<div class='svg-btn fa fa-" + icon + "' title='" + title + "'><span></span></div>");
var btnElement = document.querySelector('div.svg-operate > :last-child')
btnElement.onclick = ()=>{drawAPI.unstable.setTextContent('loading script')}

var svg = drawAPI.unstable.getSVGElement()

var getStrokeGroups = () => {
    let strokes = []
    for (const item of svg.children) {
        if (item.nodeName === 'path') {
            let points = item
                .getAttributeNS(null, "d")
                .split(/M |L /g)
                .slice(1)
                .map(item => {
                    return {
                        x: parseFloat(item.split(",")[0]),
                        y: parseFloat(item.split(",")[1])
                    };
                })
            strokes.push({ x: points.map(v => v.x), y: points.map(v => v.y) })
        }
    }
    console.log(strokes);
    return [{ "penStyle": null, "strokes": strokes }]
}

var s1 = document.createElement('script')
s1.setAttribute('nonce', drawAPI.unstable.nonce())
s1.src = iink_cdn
s1.onload = () => {

    const onclick = () => {
        drawAPI.unstable.setTextContent('calling the API')
        // Creating a recognizer
        const iinkRecognizer = iink.DefaultBehaviors.recognizerList.find(x => {
            const infos = x.getInfo();
            return infos.protocol === 'REST';
        });

        // Creating a empty model
        const model = iink.InkModel.createModel();
        // Filling the model with the stroke groups
        model.strokeGroups = getStrokeGroups();

        // Creating a recognizer context with the configuration attached
        const recognizerContext = iink.RecognizerContext.createEmptyRecognizerContext({
            configuration: iink.DefaultConfiguration
        });

        recognizerContext.editor.configuration.recognitionParams = {
            type: 'MATH',
            protocol: 'REST',
            server: {
                scheme: 'https',
                // host: 'webdemoapi.myscript.com',
                host: 'cloud.myscript.com',
                applicationKey: token.applicationKey,
                hmacKey: token.hmacKey
            },
            iink: {
                math: {
                    mimeTypes: [
                        'application/x-latex',
                    ],
                }
            }
        }

        // Assigning a theme to the document
        recognizerContext.editor.theme = iink.DefaultTheme;

        // Defining the behaviour on recognition result
        const recognitionCallback = (err, x) => {
            if (!err) {
                Object.entries(x.exports)
                    .forEach(([mimeType, exportValue]) => {

                        let latex = x.exports[mimeType]
                        let content = '\n$$'+latex.trim()+'$$ '+' '
                        drawAPI.unstable.setTextContent('')
                        drawAPI.unstable.editCurrentLine({
                            control: 0,
                            text: content
                        })
                    });
            }
        };

        // Triggering the recognition
        iinkRecognizer.export_(recognizerContext, model)
            .then((values) => {
                values.forEach((value) => {
                    recognitionCallback(undefined, value);
                });
            })
            .catch(err => recognitionCallback(err, undefined));
    }

    btnElement.onclick = onclick
}
document.body.appendChild(s1)

}
var sss=fff.toString().slice(7,-2).replace(/ +/g,' ')
var ooo={"type":"script","version":"0.1.2","function":sss}
console.log(JSON.stringify(ooo))
// drawAPI.unstable.customOperate(ooo)

buttons_demo/ocr_to_latex_mathpix.js
/* 
Recognitize handwritten text to latex using api of mathpix
Its API it free for 1000/month, but you have to provide a card.
They charge a one-time non-refundable setup fee of $1.

Also support drag and paste image to latex
*/
var fff=()=>{

var token = { app_id: 'PasteYourTokenHere', app_key: 'PasteYourTokenHere' }
var addGap = true
var icon = 'square-root-alt'
var title = 'Recognize to latex'

document.querySelector('div.svg-operate').insertAdjacentHTML('beforeEnd', (addGap ? "<span class='svgiconspliter'></span>" : "") + "<div class='svg-btn fa fa-" + icon + "' title='" + title + "'><span></span></div>");
var btnElement = document.querySelector('div.svg-operate > :last-child')
btnElement.onclick = () => {

    drawAPI.unstable.getPNG((dataURL) => {
        drawAPI.unstable.setTextContent('calling the API')
        xhrPost(dataURL, (err,ret)=>{
            console.log(err,ret)
            let latex = JSON.parse(ret)['latex']
            let content = '\n$$'+latex.trim()+'$$ '+' '
            drawAPI.unstable.setTextContent('')
            drawAPI.unstable.editCurrentLine({
                control: 0,
                text: content
            })
        })
    })
}

function xhrPost(dataURL, callback) {
    var xhr = new XMLHttpRequest();
    xhr.onreadystatechange = function () {
        if (xhr.readyState == 4) {
            if ((xhr.status >= 200 && xhr.status < 300) || xhr.status == 304) {
                callback(null, xhr.responseText);
            } else {
                callback([xhr.status, xhr.responseText], null);
            }
        }
    }
    xhr.open('post', 'https://api.mathpix.com/v3/latex');
    xhr.setRequestHeader('app_id', token.app_id)
    xhr.setRequestHeader('app_key', token.app_key)
    xhr.setRequestHeader('Content-type', 'application/json')
    xhr.send(JSON.stringify({ 'url': dataURL }));
}

// drag and paste image
function convertImage(dataURL) {
    drawAPI.unstable.setTextContent('calling the API to convert image')
    xhrPost(dataURL, (err,ret)=>{
        console.log(err,ret)
        let latex = JSON.parse(ret)['latex']
        let content = '\n$$'+latex.trim()+'$$ '+' '
        drawAPI.unstable.setTextContent('')
        drawAPI.unstable.editCurrentLine({
            control: 0,
            text: content
        })
    })

}
function getImage(items, cb) {
    var file = null;
    if (items && items.length) {
        for (var i = 0; i < items.length; i++) {
            if (items[i].type.indexOf('image') !== -1) {
                file = items[i].getAsFile();
                break;
            }
        }
    }
    // console.log(file);
    if (file) {
        var reader = new FileReader()
        reader.onload = function (event) {
            cb(event.target.result);
        }
        reader.readAsDataURL(file);
    }
}
var bindElement = document.body
bindElement.addEventListener('paste', function (event) {
    var items = event.clipboardData?.items;
    getImage(items,convertImage);
});
bindElement.ondragover = function (ev) {
    ev.preventDefault();
}
bindElement.ondrop = function (ev) {
    ev.preventDefault();
    var items = ev.dataTransfer.items;
    getImage(items,convertImage);
}

}
// {"request_id":"xxx","error":"","latex":"\\frac { - b \\pm \\sqrt { b ^ { 2 } - 4 a c } } { 2 a }","latex_list":[],"latex_confidence":0.869140625,"latex_confidence_rate":0.9940518465909091,"detection_map":{"contains_table":0,"contains_chart":0,"contains_diagram":0,"contains_graph":0,"is_blank":0.0008327960968017578,"is_inverted":0,"is_printed":0.00012830943160224706,"is_not_math":0},"detection_list":[],"position":{"top_left_x":6,"top_left_y":3,"width":403,"height":151},"auto_rotate_confidence":0.0008321025474629096,"auto_rotate_degrees":0}

var sss=fff.toString().slice(7,-2).replace(/ +/g,' ')
var ooo={"type":"script","version":"0.1.2","function":sss}
console.log(JSON.stringify(ooo))
// drawAPI.unstable.customOperate(ooo)